prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt

```diff
@@ -183,21 +198,139 @@ class FHIRRouter(
         db.insertTask(report, reportFormat, reportUrl, nextAction, null)
     }
 
+    /**
+     * Applies all filters to the list of all receivers with topic FULL_ELR that are not set as INACTIVE.
+     * FHIRPath expressions are run against the [bundle] to determine if the receiver should get this message\
+     * @return list of receivers that should receive this bundle
+     */
+    internal fun applyFilters(bundle: Bundle): List<Receiver> {
+        val listOfReceivers = mutableListOf<Receiver>()
+        // find all receivers that have the full ELR topic and determine which applies
+        val fullElrReceivers = settings.receivers.filter {
+            it.customerStatus != CustomerStatus.INACTIVE &&
+                it.topic == Topic.FULL_ELR.json_val
+        }
+
+        // get the quality filter default result for the bundle, but only if it is needed
+        val qualFilterDefaultResult: Boolean by lazy {
+            evaluateFilterCondition(
+                qualityFilterDefault, bundle,
+                false
+            )
+        }
+        // get the processing mode (processing id) default result for the bundle, but only if it is needed
+        val processingModeDefaultResult: Boolean by lazy {
+            evaluateFilterCondition(
+                processingModeFilterDefault, bundle,
+                false
+            )
+        }
+
+        fullElrReceivers.forEach { receiver ->
+            // get the receiver's organization, since we need to be able to find/combine the correct filters
+            val orgFilters = settings.findOrganization(receiver.organizationName)!!.filters
```

Comment:
(Question) Hadn't heard of `!!` before seeing this, why do we want a null pointer exception here in the case where we don't find an organization? How is that surfaced?

Reply:
settings.findOrganization(<name>) is used in a number of places. In this case we are using a defined name that we know will not be null, so use !! to indicate to kotlin that it is not nullable even though the return type is. This same underlying method is used during authentication to look for the organization contained within the client header of the request. Since that is user-provided input it may return null.

Reply:
Ah gotcha, so basically `!!` is used to label something as non-nullable, and if it turns out it is null it'll throw an exception, but the exception is not the reason for using `!!` generally, the exception is just kind of a necessary part of handling it that way. And we can't change the method return type to be non-null, because there are other cases where it should be null. I think I understand. Even knowing it'll not be null, I could see us adding a quick catch as a safety net, but if you're not worried about it, then that's probably overkill.